crates/hir/src/diagnostics.rs

```diff
@@ -444,6 +451,20 @@ impl AnyDiagnostic {
                     Err(SyntheticSyntax) => (),
                 }
             }
+            BodyValidationDiagnostic::RemoveTrailingReturn { return_expr } => {
+                if let Ok(source_ptr) = source_map.expr_syntax(return_expr) {
+                    // Filters out desugared return expressions (e.g. desugared try operators).
+                    if ast::ReturnExpr::can_cast(source_ptr.value.kind()) {
```

Comment:
```suggestion
                    if let Some(ptr) = source_ptr.value.cast::<ast::ReturnExpr>()) {
```
Then we can put a `Astptr<ast::ReturnExpr>` in the diagnostic and handle that properly